posts/2022-07-25-keyword-generics.md

```diff
@@ -0,0 +1,343 @@
+---
+layout: post
+title: "Announcing the Keyword Generics Initiative"
+author: Yoshua Wuyts, on behalf of the Keyword Generics Initiative
+release: false
+---
+
+We ([Oli], [Niko], and [Yosh]) are excited to announce the start of the [Keyword
+Generics Initiative][kwi], a new initiative [^initiative] under the purview of
+the language team. We're officially just a few weeks old now, and in this post
+we want to briefly share why we've started this initiative, and share some
+insight on what we're about.
+
+[Oli]: https://github.com/oli-obk
+[Niko]: https://github.com/nikomatsakis
+[Yosh]: https://github.com/yoshuawuyts
+[kwi]: https://github.com/rust-lang/keyword-generics-initiative
+
+[^initiative]: Rust governance terminology can sometimes get confusing. An
+"initiative" in Rust parlance is different from a "working group" or "team".
+Initiatives are intentionally limited: they exist to explore, design, and
+implement specific pieces of work - and once that work comes to a close, the
+initiative will wind back down. This is different from, say, the lang team -
+which essentially carries a `'static` lifetime - and whose work (ominously) does
+not have a clearly defined end.
+
+## A missing kind of generic
+
+One of Rust's defining features is the ability to write functions which are
+_generic_ over their input types. That allows us to write a function once,
+leaving it up to the compiler to generate the right implementations for us.
+
+But while we're able to be generic over _types_ in a function, we're not able to
+be generic over the qualifier keywords of functions and traits. The post ["What
+color is your function"][color] [^color] describes what happens when you're not
+able to be generic over the `async` qualifier. But we see this problem apply for
+practically all qualifier keywords. We're looking to fill in that gap in our
+capabilities by introducing "keyword generics" [^name]: the ability to be
+generic over keywords such as `const` and `async`.
+
+[color]: https://journal.stuffwithstuff.com/2015/02/01/what-color-is-your-function/
+[^color]: R. Nystrom, “What Color is Your Function?,” Feb. 01, 2015.
+https://journal.stuffwithstuff.com/2015/02/01/what-color-is-your-function/
+(accessed Apr. 06, 2022).
+
+[^name]: The longer, more specific name would be: "keyword modifier generics".
+We've tried calling it that, but it's a bit of a mouthful. So we're just
+sticking with "keyword generics" for now, even if the name for this feature may
+end up being called something more specific in the reference and documentation.
+
+To give you a quick taste of what we're working on, this is how we imagine you
```

Comment:
I would explicitly note that this syntax is just for demonstration is not really the point. The point is what it allows one to express.

Reply:
Good call!